docs/user-guide/setting-up-ci-cd.rst

```diff
@@ -0,0 +1,212 @@
+new problems or negatively impact the performance of your assistant. Most tests are
+quick enough to run on every change. However, you can set some more
+resource-intentsive tests to run only when certain files have been changed or
+when a certain tag is present.
+
+CI checks usually run on commit, or on merge/pull request.
+
+.. contents::
+   :local:
+
+Validate Data
+#############
+
+:ref:`Data validation <validate-files>` verifies that there are no mistakes or
+major inconsistencies in your domain file, NLU data, or story data. 
+
+.. code-block:: bash
+
+   rasa data validate --fail-on-warnings
+
+If ``rasa
+data validate`` results in errors, training a model will also fail. By
+including the ``--fail-on-warnings`` flag, the command will also fail on
+warnings about problems that won't prevent training a model, but might indicate
+messy data, such as actions listed in the domain that aren't used in any
+stories.
+
+Validate Stories
+################
+
+:ref:`Story validation <test-story-files-for-conflicts>` checks if you have any
+stories where different bot actions follow from the same dialogue history.
+
+.. code-block:: bash
+
+   rasa data validate --fail-on-warnings --max-history 5
+
+Conflicts between stories will prevent Rasa from learning the correct pattern.
+You can run story validation by passing the ``--max-history`` flag to ``rasa
+data validate``, either in a seperate check or as part of the data validation
+check. ``--max-history`` should be set to the value of ``max_history`` in your
+config for your memoization policy (which defaults to ``5``).
+
+Train a Model
+#############
+
+.. code-block:: bash
+
+   rasa train
+
+Training a model verifies that your NLU pipeline and policy configurations are
+valid and trainable, and it provides a model to use for test conversations. Training a model is also :ref:`part of the continuous deployment
+process <uploading-a-model>`, as you'll need a model to upload to your server. 
+
+Test the Assistant
+##################
+
+Testing your trained model on :ref:`test conversations
+<end-to-end-testing>` is the best way to have confidence in how your assistant
+will act in certain situations. These stories, written in a modified story
+format, allow you to provide entire conversations and test that, given this
+user input, your model will behave in the expected manner. This is especially
+important as you start introducing more complicated stories from user
+conversations. End-to-end testing is only as thorough and accurate as the test
+cases you write, so you should always update your test conversations
+whenever you make changes to your training stories.
+
+.. code-block:: bash
+
+   rasa test --stories tests/test_stories.md --fail-on-prediction-errors
+
+To ensure the test will fail if any test conversation fails, add 
+the ``--fail-on-prediction-errors`` flag:
+
+Note: End-to-end testing does **not** execute your action code. You will need to
```

Comment:
I think I would add a section about turning real conversations into end-to-end tests using Rasa X. Maybe a quick section on how it works and then by a rule of thumb you want to roughly collect a sample of the true distribution of real conversations to use as tests. They get automatically added to tests/conversation_tests.md and then will run them as part of an automated testing pipeline.

Reply:
I think that should live in the rasa x CI/CD docs, not here

Reply:
> by a rule of thumb you want to roughly collect a sample of the true distribution of real conversations to use as tests

this is a valid addition though

Reply:
Ok, I'll add something in the Rasa X pr https://github.com/RasaHQ/rasa-x/pull/2413

Reply:
Oh actually we can just link to the part about adding e2e tests from convos in improving assistants

Reply:
I did not realize there was a Rasa X CI/CD docs PR. That makes sense.

Maybe "by a rule of thumb you want to roughly collect a sample of the true distribution of real conversations to use as tests" + a great way to do that is with Rasa X (linked to Rasa X docs) is the way to do it

Reply:
> Oh actually we can just link to the part about adding e2e tests from convos in improving assistants

I tried to do that earlier and it does not have its own header, so you can't link to it directly unfortunately. We should have something about this in the Rasa X CI/CD docs too, I think

Reply:
I'll make a header for it, that's a long section anyway